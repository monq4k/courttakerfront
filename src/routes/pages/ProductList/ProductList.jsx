import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';

import { Grid, Typography, Box } from '@material-ui/core';
import { RemoveShoppingCart, StarBorder } from '@material-ui/icons';

import ProductSlider from './components/ProductSlider';
import ProductItemConnect from './components/ProductItem';
import InfinityScroll from './components/InfinityScroll';
import ProductListBar from './components/ProductListBar';
import EmptyListReplacer from '../../../shared/EmptyListReplacer';

import useStyles from './ProductListStyles';

const ProductList = ({
  history,
  cardList,
  requestAllProducts,
  totalCount,
  favorite,
  cart,
}) => {
  const classes = useStyles();

  const [isShowFavorite, setIsShowFavorite] = useState(false);
  const [isShowInCart, setIsShowInCart] = useState(false);
  const [searchValue, setSearchValue] = useState('');

  const [page, setPage] = useState(1);
  const fetchOptions = {
    page,
    limit: 8,
    title_like: searchValue,
  };

  const [contentToRender, setContentToRender] = useState([]);

  const initialGetDataRequest = (title_like = '') => {
    setPage(1);
    requestAllProducts({ ...fetchOptions, page: 1, title_like }).then(
      () => {
        setPage(2);
      },
    );
  };

  useEffect(() => {
    setContentToRender(cardList);
  }, [cardList]);

  return (
    <>
      <ProductSlider
        items={cardList.filter((card) => card.id < 8).map((card) => card.img)}
      />
      <ProductListBar
        initialGetDataRequest={initialGetDataRequest}
        setContentToRender={setContentToRender}
        isShowFavorite={isShowFavorite}
        isShowInCart={isShowInCart}
        setIsShowFavorite={setIsShowFavorite}
        setIsShowInCart={setIsShowInCart}
        searchValue={searchValue}
        setSearchValue={setSearchValue}
        favorite={favorite}
        cart={cart}
      />
      <Grid container spacing={3} className={classes.root}>
        {(isShowInCart && isShowFavorite && cart.length === 0 && favorite.length === 0 && (
          <EmptyListReplacer>
            <Box className={classes.emptyFilterContainer}>
              <RemoveShoppingCart color="disabled" className={classes.emptyFilterIcon} />
              <StarBorder color="disabled" className={classes.emptyFilterIcon} />
              <Typography color="secondary" className={classes.emptyFilterText}>
                There are no products in cart and in favorite!
              </Typography>
            </Box>
          </EmptyListReplacer>
        ))
        || (isShowFavorite && favorite.length === 0 && (
          <EmptyListReplacer>
            <Box className={classes.emptyFilterContainer}>
              <StarBorder color="disabled" className={classes.emptyFilterIcon} />
              <Typography color="secondary" className={classes.emptyFilterText}>
                There are no products in favorite!
              </Typography>
            </Box>
          </EmptyListReplacer>
        ))
        || (isShowInCart && cart.length === 0 && (
          <EmptyListReplacer>
            <Box className={classes.emptyFilterContainer}>
              <RemoveShoppingCart color="disabled" className={classes.emptyFilterIcon} />
              <Typography color="secondary" className={classes.emptyFilterText}>
                There are no products in cart!
              </Typography>
            </Box>
          </EmptyListReplacer>
        ))
        || (contentToRender.map((card) => (
          <ProductItemConnect
            isInFavorite={!!favorite.find((product) => product.id === card.id)}
            isInCart={!!cart.find((product) => product.id === card.id)}
            key={card.id}
            history={history}
            card={card}
          />
        )))}
      </Grid>
      {!isShowFavorite && !isShowInCart && (
        <InfinityScroll
          page={page}
          pageLimit={Math.ceil(totalCount / 8)}
          fetchOptions={fetchOptions}
          setPage={setPage}
          requestAllProducts={requestAllProducts}
        />
      )}
    </>
  );
};

ProductList.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,

  cardList: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      img: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired,
    }),
  ).isRequired,

  totalCount: PropTypes.number.isRequired,

  requestAllProducts: PropTypes.func.isRequired,

  favorite: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      img: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired,
    }),
  ).isRequired,
  cart: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      img: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired,
    }),
  ).isRequired,
};

export default withRouter(ProductList);
