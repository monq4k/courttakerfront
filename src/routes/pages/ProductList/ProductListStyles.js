import { makeStyles } from '@material-ui/core';

const useStyles = makeStyles({
  root: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
  emptyFilterContainer: {
    textAlign: 'center',
    marginTop: '50px',
  },
  emptyFilterIcon: {
    fontSize: '100px',
  },
  emptyFilterText: {
    fontSize: '18px',
    fontWeight: 'bold',
    textAlign: 'center',
  },
});

export default useStyles;
