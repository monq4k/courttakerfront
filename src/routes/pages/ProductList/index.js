import ProductList from './ProductList';
import ProductListContainer from './ProductListContainer';

const ProductListConnect = ProductListContainer(ProductList);

export default ProductListConnect;
