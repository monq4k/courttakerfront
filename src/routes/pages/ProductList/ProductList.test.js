import React from 'react';
import { shallow } from 'enzyme';

import ProductList from './ProductList';

const products = [
  {
    id: 0,
    title: 'Ergonomic Fresh Bacon',
    img: 'https://unsplash.it/720/1280?image=957',
    description:
      'The slim & simple Maple Gaming Keyboard from Dev Byte comes with a sleek body and 7- Color RGB LED Back-lighting for smart functionality',
    price: 209,
  },
  {
    id: 1,
    title: 'Fantastic Metal Gloves',
    img: 'https://unsplash.it/720/1280?image=221',
    description:
      'The beautiful range of Apple Naturalé that has an exciting mix of natural ingredients. With the Goodness of 100% Natural Ingredients',
    price: 634,
  },
];

const defaultProps = {
  cardList: [],
  requestAllProducts: jest.fn(),
};
let wrapper;

beforeEach(() => {
  jest.clearAllMocks();
  jest.spyOn(React, 'useEffect').mockImplementation((f) => f());
  wrapper = shallow(<ProductList {...defaultProps} />);
});

describe('ProductList', () => {
  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('cardList prop should be not null', () => {
    expect(wrapper.prop('cardList')).not.toBeNull();
  });

  it('contentToRender should contains only favorite after click on showFavButton', () => {
    const props = {
      cardList: products,
      favorite: [{
        id: 1,
        title: 'Fantastic Metal Gloves',
        img: 'https://unsplash.it/720/1280?image=221',
        description:
          'The beautiful range of Apple Naturalé that has an exciting mix of natural ingredients. With the Goodness of 100% Natural Ingredients',
        price: 634,
      }],
    };
    wrapper = shallow(<ProductList {...props} />);
    expect(wrapper.state('isShowFavorite')).to.equal(false);
  });
  // should call requestAllProducts
  // cardList
  // instance
  // onClick filters button
  // in actions - fetch-mock
});
