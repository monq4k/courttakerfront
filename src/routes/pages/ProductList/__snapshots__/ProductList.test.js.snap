// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList should render correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
